### PR TITLE
Add Support for Storage Operations

### PR DESCRIPTION
**Pull request by Ellerbach**
## Description

- Implement Support for Storage Operation.
- Add implementation for ESP32 using internal storage.
- Add implementation for ChibiOS (STM32) using SPIFFs.

## Motivation and Context

- Add Support for Storage Operation
- Implemented for ESP32 only on internal storage
- Supports deletion of a file and writing a file
- Following nanoframework/nf-debugger#374
- Addresses nanoframework/Home#1452

## How Has This Been Tested?

- Tested on a real ESP32.
- Tested on a ORGPAL3 target.

## Types of changes

- [ ] Improvement (non-breaking change that improves a feature, code or algorithm)
- [ ] Bug fix (non-breaking change which fixes an issue with code or algorithm)
- [x] New feature (non-breaking change which adds functionality to code)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] Config and build (change in the configuration and build system, has no impact on code or features)
- [ ] Dev Containers (changes related with Dev Containers, has no impact on code or features)
- [ ] Dependencies (update dependencies and changes associated, has no impact on code or features)
- [ ] Documentation (changes or updates in the documentation, has no impact on code or features)

## Checklist

- [x] My code follows the code style of this project (only if there are changes in source code).
- [ ] My changes require an update to the documentation (there are changes that require the docs website to be updated).
- [ ] I have updated the documentation accordingly (the changes require an update on the docs in this repo).
- [x] I have read the [CONTRIBUTING](https://github.com/nanoframework/.github/blob/main/CONTRIBUTING.md) document.
- [x] I have tested everything locally and all new and existing tests passed (only if there are changes in source code).

### CODE REVIEW

**Comment by networkfusion** [comment]
Is their any reason why this a Hardware Abstraction rather than a Platform Abstraction?

File: CMakeLists.txt [hunk 1]
```diff
@@ -287,6 +287,19 @@ else()
     message(STATUS "Configuration block storage **IS NOT** included")
 endif()
 
+#################################################################
+# enables internal storage support
+# (default is OFF so Configuration block storage is NOT supported)
```
Comment (josesimoes) [inline CMakeLists.txt:292]:
```suggestion
# (default is OFF so internal storage is NOT supported)
```

File: src/HAL/Include/nanoHAL_StorageOperation.h [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+        uint32_t nameLength,
+        uint32_t dataLength,
+        uint8_t *data,
+        uint32_t *errorCode);
+
+#ifdef __cplusplus
+}
+#endif
+
+#endif // NANOHAL_STORAGE_OPERATIN_H
```
Comment (josesimoes) [inline src/HAL/Include/nanoHAL_StorageOperation.h:27]:
```suggestion
#endif // NANOHAL_STORAGE_OPERATION_H

```

File: src/CLR/Debugger/Debugger.cpp [hunk 3]
```diff
@@ -1335,6 +1336,29 @@ bool CLR_DBG_Debugger::Monitor_UpdateConfiguration(WP_Message *message)
...
+{
+    NATIVE_PROFILE_CLR_DEBUGGER();
+
+#if (HAS_INTERNAL_STORAGE == TRUE)
+
+    Monitor_StorageOperation_Command *cmd = (Monitor_StorageOperation_Command *)message->m_payload;
+    Monitor_StorageOperation_Reply cmdReply;
+    uint32_t err = 0;
+
+    HAL_StorageOperation(cmd->Operation, cmd->NameLength, cmd->DataLength, cmd->Data, &err);
```
Comment (josesimoes) [inline src/CLR/Debugger/Debugger.cpp:1349]: This has to handle WP chunked transfers...

**Review by josesimoes — COMMENTED** [review]
Nice! Still some work required, but overall, that's it.

**Comment by Ellerbach** [comment]
> Is their any reason why this a Hardware Abstraction rather than a Platform Abstraction?

I mimic the Configuration Manager which was HAL. I guess it does not have much impact.

**Comment by Ellerbach** [comment]
Good to go! Thanks for the improvements on my initial code!

**Comment by josesimoes** [comment]
I'm adding implementation for this to ORGPAL target before merging.

**Review by josesimoes — APPROVED** [review]

File: targets/ChibiOS/_include/targetHAL_StorageOperation.h [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+//
+
+#ifndef TARGET_PAL_STORAGE_OPERATION_H
+#define TARGET_PAL_STORAGE_OPERATION_H
+
+#include <WireProtocol_MonitorCommands.h>
+#include <Debugger.h>
+#include <nanoHAL_StorageOperation.h>
+
+#endif  // TARGET_PAL_STORAGE_OPERATION_H
```
Comment (networkfusion) [inline targets/ChibiOS/_include/targetHAL_StorageOperation.h:13]: filename is HAL, but content is PAL

**Pull request merged**